### ISSUE TITLE
sp_BlitzIndex: old statistics query slow with > 50000 objects

### ISSUE THREAD

**Issue by BlitzErik** [BrentOzarULTD/SQL-Server-First-Responder-Kit#527]
**Do you want to request a *feature* or report a *bug*?**
Improvement/performance bug.

**What is the current behavior?**
When using the old syntax that hits sys.sysindexes, the statistics query performs very badly when you have > 50,000 statistics objects to analyze.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via http://sqlfiddle.com

Create a ton of statistics:

```
SELECT 'CREATE STATISTICS [IX_' + 
		c1.COLUMN_NAME + '_' + CONVERT(VARCHAR(36), NEWID()) + ']'
    + ' ON [' + c1.TABLE_SCHEMA + '].[' + c1.TABLE_NAME + '] ' 
    + '([' + c1.COLUMN_NAME + '])' 
FROM INFORMATION_SCHEMA.COLUMNS c1
JOIN INFORMATION_SCHEMA.TABLES AS t
ON c1.TABLE_NAME = t.TABLE_NAME
CROSS JOIN (SELECT TOP 10 * FROM sys.messages) AS m --change the top to a higher number to create more
WHERE c1.TABLE_SCHEMA = 'dbo' 
AND (c1.CHARACTER_MAXIMUM_LENGTH <> -1 OR c1.CHARACTER_MAXIMUM_LENGTH IS NULL)
AND t.TABLE_TYPE <> 'VIEW'
ORDER BY c1.TABLE_NAME, c1.ORDINAL_POSITION
```

Run old syntax query to gather information:

```
SELECT  DB_NAME() AS DatabaseName,
		OBJECT_NAME(s.object_id) AS table_name,
		SCHEMA_NAME(obj.schema_id) AS schema_name,
        ISNULL(i.name, 'System Statistic') AS index_name,
        c.name AS column_name,
        s.name AS statistics_name,
        CONVERT(DATE, STATS_DATE(s.object_id, s.stats_id)) AS last_statistics_update,
        DATEDIFF(DAY, STATS_DATE(s.object_id, s.stats_id), SYSDATETIME()) AS days_since_last_stats_update,
		si.rowcnt,
		si.rows,
        si.rowmodctr,
        CASE WHEN si.rowmodctr > 0 THEN CAST(si.rowmodctr / ( 1. * si.rowcnt ) * 100 AS DECIMAL(18, 1))
             ELSE si.rowmodctr
        END AS percent_modifications,
        CASE WHEN si.rowcnt < 500 THEN 500
             ELSE CAST(( si.rowcnt * .20 ) + 500 AS INT)
        END AS modifications_before_auto_update,
        ISNULL(i.type_desc, 'System Statistic - N/A') AS index_type_desc,
        CONVERT(DATE, obj.create_date) AS table_create_date,
        CONVERT(DATE, obj.modify_date) AS table_modify_date,
		s.no_recompute,
		s.has_filter,
		s.filter_definition
FROM    sys.stats AS s
LEFT JOIN    sys.stats_columns sc
ON      sc.object_id = s.object_id
        AND sc.stats_id = s.stats_id
LEFT JOIN    sys.columns c
ON      c.object_id = sc.object_id
        AND c.column_id = sc.column_id
LEFT JOIN    sys.objects obj
ON      s.object_id = obj.object_id
LEFT JOIN    sys.indexes AS i
ON      i.object_id = s.object_id
        AND i.index_id = s.stats_id
LEFT  JOIN sys.sysindexes AS si
ON si.name = s.name
WHERE obj.is_ms_shipped = 0
```

Note: the 'new' syntax query runs in 2 seconds here, and I usually kill the old syntax off around the 30 second mark. Root cause: impatience.

If you add

`AND si.rowcnt >= 1`

To the WHERE clause, it finishes just about instantly. 

```
 SQL Server Execution Times:
   CPU time = 828 ms,  elapsed time = 1017 ms.
```

**What is the expected behavior?**
Run fast.

**Which versions of SQL Server and which OS are affected by this issue? Did this work in previous versions of our procedures?**
2005, 2008, 2008R2 prior to SP2, 2012 prior to SP1

**Fix:
The fix for the old syntax is to filter out statistics that have >=1 row. This has a side effect: it filters out all system statistics. I can't seem to find a way to get sys.sysindexes to register system stats as having rows. I've been checking this over in the lab on 2005-2008R2. It's pretty weird.

```
CREATE TABLE dbo.WhatStats (col1 INT, col2 INT, col3 INT, col4 INT, col5 INT)

INSERT dbo.WhatStats WITH (TABLOCK)
SELECT N, N, N, N, N
FROM (
SELECT ROW_NUMBER() OVER (ORDER BY (SELECT NULL)) AS N
FROM sys.messages
) x

select COUNT(*)
FROM dbo.WhatStats
WHERE col1 > 0
OR col2 > 0
OR col3 > 0
OR col4 > 0
OR col5 > 0

UPDATE dbo.WhatStats
SET 
col1 = col2 * 2,
col2 = col3 * 2,
col3 = col4 * 2,
col4 = col5 * 2,
col5 = col1 * 2

```

After count:

![image](https://cloud.githubusercontent.com/assets/11590415/19410683/64ef3338-92c0-11e6-8b9b-319c54e06914.png)

After update:

![image](https://cloud.githubusercontent.com/assets/11590415/19410710/007121f4-92c1-11e6-9bfa-8a826a3a3d9c.png)

**Other stuff I tried**

Filtering on rowmodctr: just as terribly slow (true even if it's a second predicate)
Preaggregating data: sys.sysindexes was still a bottleneck
Inserting base set of data to temp table and updating: sys.sysindexes was still a bottleneck
Other query rewrites: cross/outer apply, CTEs

**Comment by BlitzErik** [comment]
#524

**Comment by BlitzErik** [comment]
Closed #524
